contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenSource.sol

```diff
@@ -0,0 +1,152 @@
+    ITeleporterReceiver,
+    INativeTokenSource,
+    ReentrancyGuard
+{
+    using SafeERC20 for IERC20;
+
+    uint256 public constant MINT_NATIVE_TOKENS_REQUIRED_GAS = 150_000; // TODO this is a placeholder
+    bytes32 public immutable currentBlockchainID;
+    bytes32 public immutable destinationBlockchainID;
+    address public immutable destinationContractAddress;
+
+    // Used for sending an receiving Teleporter messages.
+    ITeleporterMessenger public immutable teleporterMessenger;
+
+    constructor(
+        address teleporterMessengerAddress,
+        bytes32 destinationBlockchainID_,
+        address destinationContractAddress_
+    ) {
+        currentBlockchainID = WarpMessenger(WARP_PRECOMPILE_ADDRESS)
+            .getBlockchainID();
+
+        require(
+            teleporterMessengerAddress != address(0),
+            "Invalid Teleporter Messenger Address"
+        );
+        teleporterMessenger = ITeleporterMessenger(teleporterMessengerAddress);
+
+        require(
+            destinationBlockchainID_ != bytes32(0),
+            "Invalid Destination Chain ID"
+        );
+        require(
+            destinationBlockchainID_ != currentBlockchainID,
+            "Cannot Bridge With Same Blockchain"
+        );
+        destinationBlockchainID = destinationBlockchainID_;
+
+        require(
+            destinationContractAddress_ != address(0),
+            "Invalid Destination Contract Address"
+        );
+        destinationContractAddress = destinationContractAddress_;
+    }
+
+    /**
+     * @dev See {ITeleporterReceiver-receiveTeleporterMessage}.
+     *
+     * Receives a Teleporter message and routes to the appropriate internal function call.
+     */
+    function receiveTeleporterMessage(
+        bytes32 senderBlockchainID,
+        address senderAddress,
+        bytes calldata message
+    ) external nonReentrant {
+        // Only allow the Teleporter messenger to deliver messages.
+        require(
+            msg.sender == address(teleporterMessenger),
+            "Unauthorized teleporter contract"
+        );
+
+        // Only allow messages from the destination chain.
+        require(
+            senderBlockchainID == destinationBlockchainID,
+            "Invalid Destination Chain"
+        );
+
+        // Only allow the partner contract to send messages.
+        require(
+            senderAddress == destinationContractAddress,
+            "Unauthorized Sender"
+        );
+
+        (address recipient, uint256 amount) = abi.decode(
+            message,
+            (address, uint256)
+        );
+        require(recipient != address(0), "Invalid Recipient Address");
+
+        // Send to recipient
+        payable(recipient).transfer(amount);
+
+        emit UnlockTokens(recipient, amount);
+    }
+
+    /**
+     * @dev See {IERC20Bridge-bridgeTokens}.
+     */
+    function transferToDestination(
+        address recipient,
+        address feeContractAddress,
+        uint256 feeAmount
```

Comment:
if we want as close to "production ready" should we add allowed relayer addresses as parameter?